Epsilon/Components/KpiMatrixComponent.cs

```diff
@@ -0,0 +1,192 @@
+
+        // Add the outcome rows.
+        await foreach (var outcomeCriterion in listOfCriteria)
+        {
+            var row = new TableRow();
+            var outcome = Outcomes.FirstOrDefault(o => o.Id == outcomeCriterion.Id);
+
+            // Create a new paragraph for outcome.Name
+            if (outcome != null)
+            {
+                var paragraphForOutcomeName = new Paragraph(new Run(new Text(outcome.Name)))
+                {
+                    ParagraphProperties = new ParagraphProperties { Justification = new Justification { Val = JustificationValues.Center, }, },
+                };
+
+                row.AppendChild(CreateTableCellWithBorders("2500", paragraphForOutcomeName));
+            }
+
+            await foreach (var sub in Submissions)
+            {
+                var cell = CreateTableCellWithBorders("100");
+                var criteria = sub.Criteria.FirstOrDefault(c => c.Id == outcome?.Id);
+                var result = sub.Results.FirstOrDefault(r => r.Outcome.Id == outcome?.Id);
+
+                var status = GetStatus(result?.Grade, criteria?.MasteryPoints);
+                var fillColor = GetColor(status);
+                cell.FirstChild?.Append(new Shading { Fill = fillColor, });
+
+                var text = result != null ? result.Outcome.Value.ShortName : "";
+                var paragraph = new Paragraph();
+                var run = new Run(new Text(text));
+                paragraph.Append(run);
+                paragraph.ParagraphProperties = new ParagraphProperties()
+                {
+                    Justification = new Justification() { Val = JustificationValues.Center, },
+                };
+
+                cell.Append(paragraph);
+                row.AppendChild(cell);
+            }
+
+            table.AppendChild(row);
+        }
+        
+        body.Append(new Paragraph(new Run(new Text(""))));
+        body.AppendChild(table);
+
+        return mainDocumentPart.Document.AppendChild(body);
+    }
+    
+    private static OutcomeGradeStatus GetStatus(double? grade, double? masteryPoints)
+    {
+        if (grade.HasValue && masteryPoints.HasValue)
+        {
+            return grade.Value >= masteryPoints.Value ? OutcomeGradeStatus.Mastered : OutcomeGradeStatus.NotMastered;
+        }
+        if (!grade.HasValue && masteryPoints.HasValue)
+        {
+            return OutcomeGradeStatus.NotAssessed;
+        }
+        return OutcomeGradeStatus.NotGraded;
+    }
+
+    private static string GetColor(OutcomeGradeStatus status)
+    {
+        return status switch
+        {
+            OutcomeGradeStatus.Mastered => "#44F656",
+            OutcomeGradeStatus.NotMastered => "#FA1818",
+            OutcomeGradeStatus.NotAssessed => "#9F2B68",
+            OutcomeGradeStatus.NotGraded => "",
+            _ => "",
+        };
+    }
+    
+    private static TableCell CreateTableCellWithBorders(string? width, params OpenXmlElement[] elements)
```

Comment:
Please implement [this function](https://github.com/tpcly/epsilon-lti/blob/9d36efddbcd0b127b05ffc76a87fd60b9ac6f21e/Epsilon/Components/CompetenceProfileComponent.cs#L280) already in this class. Instead of this one.